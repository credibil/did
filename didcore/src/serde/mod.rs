pub mod flexvec_or_single;
pub mod option_flexvec;
pub mod option_flexvec_or_single;
