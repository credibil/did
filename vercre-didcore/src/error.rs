//! # DID Core Errors
//!
//! This module defines the error types used by the DID Core library, including for traits that
//! may be implemented in other crates.

use std::fmt::Display;

use thiserror::Error;

/// Simplify creation of errors with tracing.
///
/// # Example
/// ```
/// use didcore::error::Err;
/// use didcore::{error, Result};
///
/// fn with_msg() -> Result<(), Err> {
///     error!(Err::InvalidRequest, "message: {}", "some message")
/// }
///
/// fn no_msg() -> Result<(), Err> {
///     error!(Err::InvalidRequest)
/// }
/// ```
#[macro_export]
macro_rules! tracerr {
    // with context
    ($code:expr, $($msg:tt)*) => {
        {
        use $crate::error::Context as _;
        tracing::error!($($msg)*);
        return Err($code).context(format!($($msg)*));
        }
    };
    // no context
    ($code:expr) => {
        {
        tracing::error!("{}", $code);
        return Err($code.into());
        }
    }
}

/// Public error type for DID Core.
#[derive(Error, Debug)]
#[error(transparent)]
pub struct Error(#[from] anyhow::Error);

impl Error {
    /// Transfer the error to `OAuth2` compatible format.
    #[must_use]
    pub fn to_json(&self) -> serde_json::Value {
        serde_json::json!({
            "error": self.0.root_cause().to_string(),
            "error_description": self.to_string(),
        })
    }

    /// Returns true if `E` is the type held by this error object.
    #[must_use]
    pub fn is(&self, err: Err) -> bool {
        self.0.downcast_ref::<Err>().map_or(false, |e| e == &err)
    }
}

/// Typed errors for DID Core.
#[derive(Clone, Copy, Error, Debug, PartialEq, Eq)]
pub enum Err {
    /// Hash is not a valid SHA-256 hash.
    #[error("invalid_hash")]
    InvalidHash,

    /// Invalid format. (See context for details)
    #[error("invalid_format")]
    InvalidFormat,

    /// Invalid input. Used where a verification fails that is more complex than a simple incorrect
    /// format. (See context for details)
    #[error("invalid_input")]
    InvalidInput,

    /// Invalid key is where the format of the key is incorrect or the cryptographic algorithm
    /// specified by the key is not supported.
    #[error("invalid_key")]
    InvalidKey,

    /// Invalid patch. This is used when a patch for a DID document is verified for consistency.
    #[error("invalid_patch")]
    InvalidPatch,

    /// Key not found. This is in response to parsing a DID Document for a public key and not
    /// finding one for the specified purpose or asking a keyring for a public key and not finding
    /// one for the specified operation.
    #[error("key_not_found")]
    KeyNotFound,

    /// An error was returned from a downstream API
    #[error("api_error")]
    ApiError,

    /// An error occurred trying to deserialize data.
    #[error("deserialization_error")]
    DeserializationError,

    /// An error occurred trying to serialize data.
    #[error("serialization_error")]
    SerializationError,

    /// Environment configuration could not be resolved.
    #[error("invalid_config")]
    InvalidConfig,

    /// Authentication failed.
    #[error("auth_error")]
    AuthError,

    /// Request failed. This is used when a request to a downstream API fails to connect or get a
    /// response.
    #[error("request_error")]
    RequestError,

    /// Failure to sign a message.
    #[error("signing_error")]
    SigningError,

    /// Failure to verify a signature.
    #[error("failed_signature_verification")]
    FailedSignatureVerification,

    /// A requested key signing algorithm is not supported by either the signer or the key store.
    #[error("unsupported_algorithm")]
    UnsupportedAlgorithm,

    /// An expiry date is in the past.
    #[error("expired")]
    Expired,

    /// An unspecified error occurred (see context for information)
    #[error("unknown")]
    Unknown,

    /// Feature is not yet implemented.
    #[error("not_implemented")]
    NotImplemented,

    /// Feature is not supported.
    #[error("not_supported")]
    NotSupported,

    /// No DID document was found for the requested DID.
    #[error("not_found")]
    NotFound,
}

/// Context is used to decorate errors with useful context information.
pub trait Context<T, E>
where
    E: std::error::Error + Send + Sync + 'static,
{
    /// Adds context to the error.
    ///
    /// # Arguments
    ///
    /// * `context` - The context to add to the error.
    ///
    /// # Returns
    ///
    /// Original return object or error with context appended.
    ///
    /// # Errors
    ///
    /// * Original error with context appended.
    fn context<C>(self, context: C) -> Result<T, Error>
    where
        C: Display + Send + Sync + 'static;
}

impl<T, E> Context<T, E> for core::result::Result<T, E>
where
    E: std::error::Error + Send + Sync + 'static,
{
    fn context<C>(self, context: C) -> Result<T, Error>
    where
        C: Display + Send + Sync + 'static,
    {
        match self {
            Ok(ok) => Ok(ok),
            Err(e) => Err(Error(anyhow::Error::from(e).context(context))),
        }
    }
}

impl From<Err> for Error {
    fn from(error: Err) -> Self {
        Error(error.into())
    }
}

impl From<base64ct::Error> for Error {
    fn from(err: base64ct::Error) -> Error {
        Error(err.into())
    }
}

impl From<ecdsa::Error> for Error {
    fn from(err: ecdsa::Error) -> Error {
        Error(err.into())
    }
}

impl From<multihash::Error> for Error {
    fn from(err: multihash::Error) -> Error {
        Error(err.into())
    }
}

impl From<regex::Error> for Error {
    fn from(err: regex::Error) -> Error {
        Error(err.into())
    }
}

impl From<reqwest::Error> for Error {
    fn from(err: reqwest::Error) -> Error {
        Error(err.into())
    }
}

impl From<serde_json::Error> for Error {
    fn from(err: serde_json::Error) -> Error {
        Error(err.into())
    }
}

impl From<std::string::FromUtf8Error> for Error {
    fn from(err: std::string::FromUtf8Error) -> Error {
        Error(err.into())
    }
}

impl From<url::ParseError> for Error {
    fn from(err: url::ParseError) -> Error {
        Error(err.into())
    }
}

#[cfg(test)]
mod test {
    use serde_json::json;
    use tracing::Level;
    use tracing_subscriber::FmtSubscriber;

    use super::*;
    use crate::Result;

    #[test]
    fn base_err() {
        let err: Error = Err::InvalidFormat.into();

        assert_eq!(
            err.to_json(),
            json!({"error":"invalid_format","error_description":"invalid_format"})
        );
    }

    #[test]
    fn context_err() {
        let res: Result<()> = Err(Err::InvalidFormat).context("Invalid format description");
        let err = res.expect_err("expected error");

        assert_eq!(
            err.to_json(),
            json!({"error":"invalid_format","error_description":"Invalid format description"})
        );
    }

    #[test]
    fn test_macro() {
        let subscriber = FmtSubscriber::builder().with_max_level(Level::ERROR).finish();
        tracing::subscriber::set_global_default(subscriber).expect("setting subscriber failed");

        let Err(e) = run_macro() else {
            panic!("expected error");
        };

        assert_eq!(e.to_string(), "test me");
    }

    fn run_macro() -> Result<()> {
        tracerr!(Err::InvalidFormat, "test {}", "me")
    }
}
