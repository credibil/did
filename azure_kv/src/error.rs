use serde::Deserialize;

/// Error returned from Azure Key Vault.
#[derive(Debug, Deserialize)]
pub struct ApiErrorResponse {
    /// Error details.
    pub error: ApiErrorDetail,
}

/// Error returned from Azure Key Vault.
#[derive(Debug, Deserialize)]
pub struct ApiErrorDetail {
    /// Error code.
    pub code: String,
    /// Error message.
    pub message: String,
}
