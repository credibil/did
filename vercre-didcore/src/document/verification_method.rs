//! Verification methods allow public keys to be associated with a DID.

use std::{convert::Infallible, fmt::Display, str::FromStr};

use serde::{Deserialize, Serialize};

use crate::keys::Jwk;

/// A DID document can express verification methods, such as cryptographic public keys, which can be
/// used to authenticate or authorize interactions with the DID subject or associated parties.
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase", default)]
pub struct VerificationMethod {
    /// Identifier for the verification method. The value must be a string that conforms to DID URL
    /// Syntax which can be a relative DID URL that is confined to the DID document. Relative URLs
    /// are assumed by default.
    pub id: String,
    /// The type of verification method. One that is registered in a DID specification registry.
    /// https://www.w3.org/TR/did-spec-registries/
    #[serde(rename = "type")]
    pub type_: String,
    /// Identifier for the controller of the verification method. A DID.
    pub controller: String,
    /// The public key material of the verification method, if applicable.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub public_key_jwk: Option<Jwk>,
    /// The public key material of the verification method, if applicable.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub public_key_multibase: Option<String>,
    /// Account ID for block-chain based public keys.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub blockchain_account_id: Option<String>,
}

/// Key purpose type.
#[derive(Clone, Copy, Debug, Default, Deserialize, Eq, Hash, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum KeyPurpose {
    #[default]
    /// The authentication verification relationship is used to specify how the DID subject is
    /// expected to be authenticated, for purposes such as logging into a website or engaging in
    /// any sort of challenge-response protocol.
    Authentication,
    /// The assertionMethod verification relationship is used to specify how the DID subject is
    /// expected to express claims, such as for the purposes of issuing a Verifiable Credential
    AssertionMethod,
    /// The capabilityInvocation verification relationship is used to specify a verification method
    /// that might be used by the DID subject to invoke a cryptographic capability, such as the
    /// authorization to update the DID Document.
    CapabilityInvocation,
    /// The capabilityDelegation verification relationship is used to specify a mechanism that might
    /// be used by the DID subject to delegate a cryptographic capability to another party, such as
    /// delegating the authority to access a specific HTTP API to a subordinate.
    CapabilityDelegation,
    /// The keyAgreement verification relationship is used to specify how an entity can generate
    /// encryption material in order to transmit confidential information intended for the DID
    /// subject, such as for the purposes of establishing a secure communication channel with the
    /// recipient.
    KeyAgreement,
}

impl Display for KeyPurpose {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            KeyPurpose::Authentication => write!(f, "authentication"),
            KeyPurpose::AssertionMethod => write!(f, "assertionMethod"),
            KeyPurpose::CapabilityInvocation => write!(f, "capabilityInvocation"),
            KeyPurpose::CapabilityDelegation => write!(f, "capabilityDelegation"),
            KeyPurpose::KeyAgreement => write!(f, "keyAgreement"),
        }
    }
}

/// A reference to a verification method or an embedded verification method object, as used by the
/// "authentication" and other similar fields in a [`DidDocument`].
#[derive(Clone, Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct VmRelationship {
    /// Key identifier referring to a verification method elsewhere in the DID document.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub key_id: Option<String>,
    /// Embedded verification method object in the case where the verification method is not
    /// referred to by key identifier.
    #[serde(flatten, skip_serializing_if = "Option::is_none")]
    pub verification_method: Option<VerificationMethod>,
}

/// Deserialise a verification method relationship in the case it is a string - the ID of a
/// verification method specified elsewhere in the DID document.
impl FromStr for VmRelationship {
    type Err = Infallible;

    fn from_str(id: &str) -> Result<Self, Self::Err> {
        Ok(Self {
            key_id: Some(id.to_string()),
            verification_method: None,
        })
    }
}

/// Partial Equality for verification method relationship.
impl PartialEq for VmRelationship {
    fn eq(&self, other: &Self) -> bool {
        if self.key_id.is_some() && other.key_id.is_some() {
            return self.key_id == other.key_id;
        }
        if let Some(me) = &self.verification_method {
            if let Some(them) = &other.verification_method {
                return me == them;
            }
        }
        false
    }
}
impl Eq for VmRelationship {}

/// Convert a verification method into a verification method relationship. Note that this will only
/// pick up the ID of the verification method to refer to and will *not* embed the verification
/// method itself. If your implementation uses embeded keys, build the reference manually.
impl From<&VerificationMethod> for VmRelationship {
    fn from(vm: &VerificationMethod) -> Self {
        Self {
            key_id: Some(vm.id.clone()),
            verification_method: None,
        }
    }
}

/// Serialize a verification method relationship to a string or object. If the `key_id` field is
/// set, serialize to a string, otherwise serialize to an embedded verification method.
impl Serialize for VmRelationship {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match (&self.key_id, &self.verification_method) {
            (Some(id), None) => serializer.serialize_str(id),
            (None, Some(vm)) => vm.serialize(serializer),
            _ => Err(serde::ser::Error::custom(
                "Verification method reference must be a string or object",
            )),
        }
    }
}
