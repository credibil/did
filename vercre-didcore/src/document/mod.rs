//! DID Document and its component data structures. Includes DID document patching and some
//! simple input verifications.

use serde::{Deserialize, Serialize};

use crate::document::{
    context::{context_serialization, Context},
    service::Service,
    verification_method::{KeyPurpose, VerificationMethod, VmRelationship},
};
use crate::error::Err;
use crate::serde::{option_flexvec, option_flexvec_or_single};
use crate::{tracerr, Result};

pub mod context;
pub mod patch;
pub mod service;
pub mod verification_method;

// ----------------------------------------------------------------------------
// DID Document
// ----------------------------------------------------------------------------

/// A DID is associated with a DID document that can be serialized into a representation of the DID.
/// https://www.w3.org/TR/did-core/
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", default)]
pub struct DidDocument {
    /// The DID document's unique identifier. It is a URI scheme conformant with RFC3986. The syntax
    /// conforms to the that of the DID method implementation: "did:{method}:{uri}", where the URI
    /// portion can be used by a resolver of the DID method to retrieve the DID document.
    pub id: String,
    /// The JSON-LD Context is either a string or a list containing any combination of strings
    /// and/or ordered maps.
    #[serde(rename = "@context", with = "context_serialization")]
    pub context: Vec<Context>,
    /// A DID controller is an entity that is authorized to make changes to a DID document. It is a
    /// DID or list of DIDs and is optional.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(with = "option_flexvec_or_single")]
    pub controller: Option<Vec<String>>,
    /// A DID subject can have multiple identifiers for different purposes, or at different times.
    /// The assertion that two or more DIDs (or other types of URI) refer to the same DID subject
    /// can be made using the alsoKnownAs property.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub also_known_as: Option<Vec<String>>,
    /// A set of parameters that can be used together with a process to independently verify a
    /// proof. For example, a cryptographic public key can be used as a verification method with
    /// respect to a digital signature; in such usage, it verifies that the signer possessed the
    /// associated cryptographic private key.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub verification_method: Option<Vec<VerificationMethod>>,
    /// Authentication methods - maps to one or more verification methods by ID or contains embedded
    /// verification methods. Specifies how the DID subject is authenticated for purposes such as
    /// logging into a website or engaging in challenge-response interactions.
    #[serde(skip_serializing_if = "Option::is_none", with = "option_flexvec")]
    pub authentication: Option<Vec<VmRelationship>>,
    /// Assertion methods - maps to one or more verification methods by ID or contains embedded
    /// verification methods. Specifies how the DID subject is expected to express claims such as
    /// for the purposes of issuing verifiable credentials, or for a verifier to determine the
    /// authenticity of the DID subject.
    #[serde(skip_serializing_if = "Option::is_none", with = "option_flexvec")]
    pub assertion_method: Option<Vec<VmRelationship>>,
    /// Key agreement methods - maps to one or more verification methods by ID or contains embedded
    /// verification methods. Specifies how an entity can generate encryption material to transmit
    /// confidential messages to the DID subject.
    #[serde(skip_serializing_if = "Option::is_none", with = "option_flexvec")]
    pub key_agreement: Option<Vec<VmRelationship>>,
    /// Capability invocation methods - maps to one or more verification methods by ID or contains
    /// embedded verification methods. Specifies how the DID subject can invoke a crypotgraphic
    /// capability, such as to authorize an update to the DID document.
    #[serde(skip_serializing_if = "Option::is_none", with = "option_flexvec")]
    pub capability_invocation: Option<Vec<VmRelationship>>,
    /// Capability delegation methods - maps to one or more verification methods by ID or contains
    /// embedded verification methods. Specifies how the DID subject can delegate a crypotgraphic
    /// capability to another party. For example, delegating the authority to access an online
    /// resource like an HTTP API.
    #[serde(skip_serializing_if = "Option::is_none", with = "option_flexvec")]
    pub capability_delegation: Option<Vec<VmRelationship>>,
    /// Services are used to express ways of communicating with the DID subject or associated
    /// entities. Can be any type of service the DID subject wants to advertise, including
    /// decentralized identity management services for further discovery, authentication,
    /// authorization, or interaction.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub service: Option<Vec<Service>>,
}

/// Utility methods for looking up DID document components.
impl DidDocument {
    // Get a verfication method from a verification method reference, either by looking up the ID
    // or returning the embedded verification method. If the ref contains an ID that doesn't resolve
    // or an embedded verification method doesn't exist, None is returned.
    fn vm_from_ref(&self, vm_ref: &VmRelationship) -> Option<VerificationMethod> {
        if let Some(id) = &vm_ref.key_id {
            if let Some(vms) = &self.verification_method {
                for vm in vms {
                    if &vm.id == id {
                        return Some(vm.clone());
                    }
                }
            }
        } else if let Some(vm) = &vm_ref.verification_method {
            return Some(vm.clone());
        }
        None
    }

    /// Get a key from the document by purpose. If the key is not found, returns an error.
    pub fn get_key(&self, purpose: KeyPurpose) -> Result<VerificationMethod> {
        let mut vm = Option::<VerificationMethod>::None;
        match purpose {
            KeyPurpose::Authentication => {
                if let Some(auth) = &self.authentication {
                    for a in auth {
                        vm = self.vm_from_ref(a);
                        if vm.is_some() {
                            break;
                        }
                    }
                }
            }
            KeyPurpose::AssertionMethod => {
                if let Some(assert) = &self.assertion_method {
                    for a in assert {
                        vm = self.vm_from_ref(a);
                        if vm.is_some() {
                            break;
                        }
                    }
                }
            }
            KeyPurpose::KeyAgreement => {
                if let Some(key) = &self.key_agreement {
                    for k in key {
                        vm = self.vm_from_ref(k);
                        if vm.is_some() {
                            break;
                        }
                    }
                }
            }
            KeyPurpose::CapabilityDelegation => {
                if let Some(cap) = &self.capability_delegation {
                    for c in cap {
                        vm = self.vm_from_ref(c);
                        if vm.is_some() {
                            break;
                        }
                    }
                }
            }
            KeyPurpose::CapabilityInvocation => {
                if let Some(cap) = &self.capability_invocation {
                    for c in cap {
                        vm = self.vm_from_ref(c);
                        if vm.is_some() {
                            break;
                        }
                    }
                }
            }
        };

        match vm {
            Some(v) => Ok(v),
            None => tracerr!(Err::KeyNotFound, "No key found for purpose"),
        }
    }
}

#[cfg(test)]
mod tests {
    use std::{collections::HashMap, fs::File, path::PathBuf, vec};

    use olpc_cjson::CanonicalFormatter;

    use super::*;
    use crate::document::service::ServiceEndpoint;
    use crate::keys::Jwk;

    fn public_key() -> Jwk {
        Jwk {
            kty: "EC".to_string(),
            crv: Some("secp256k1".to_string()),
            x: Some("smmFWI4qLfWztIzwurLCvjjw7guNZvN99ai2oTXGUtc".to_string()),
            y: Some("rxp_kiiXHitxLHe545cePsF0y_Mdv_dy6zY4ov_0q9g".to_string()),
            ..Default::default()
        }
    }

    fn default_service() -> Service {
        Service {
            id: "service1".to_string(),
            type_: vec!["service1type".to_string()],
            service_endpoint: vec![ServiceEndpoint {
                url: Some("https://service1.example.com/".to_string()),
                url_map: None,
            }],
        }
    }

    fn default_doc() -> DidDocument {
        DidDocument {
            id: "did:ion:EiAscM5K0lfATv8GEqlR_RAVId0alzdcOgIRs-fBLXBWFA".to_string(),
            context: vec![Context {
                url: Some(context::DID_CONTEXT.to_string()),
                url_map: None,
            }],
            controller: Some(vec![
                "did:ion:EiAscM5K0lfATv8GEqlR_RAVId0alzdcOgIRs-fBLXBWFA".to_string(),
            ]),
            also_known_as: None,
            verification_method: Some(vec![VerificationMethod {
                id: "371544b48d7d60d430c9c8b4af3745fa".to_string(),
                controller: "did:ion:EiAscM5K0lfATv8GEqlR_RAVId0alzdcOgIRs-fBLXBWFA".to_string(),
                type_: "EcdsaSecp256k1VerificationKey2019".to_string(),
                public_key_jwk: Some(public_key()),
                ..Default::default()
            }]),
            authentication: Some(vec![VmRelationship {
                key_id: Some("371544b48d7d60d430c9c8b4af3745fa".to_string()),
                verification_method: None,
            }]),
            assertion_method: Some(vec![VmRelationship {
                key_id: Some("371544b48d7d60d430c9c8b4af3745fa".to_string()),
                verification_method: None,
            }]),
            key_agreement: None,
            capability_invocation: None,
            capability_delegation: None,
            service: Some(vec![default_service()]),
        }
    }

    #[test]
    fn default_doc_is_empty() {
        let doc = DidDocument::default();
        assert_eq!(doc.id, "");
        assert!(doc.context.is_empty());
        assert!(doc.controller.is_none());
        assert!(doc.also_known_as.is_none());
        assert!(doc.verification_method.is_none());
        assert!(doc.authentication.is_none());
        assert!(doc.assertion_method.is_none());
        assert!(doc.key_agreement.is_none());
        assert!(doc.capability_invocation.is_none());
        assert!(doc.capability_delegation.is_none());
        assert!(doc.service.is_none());
    }

    #[test]
    fn serialize_constructed_doc() {
        let doc = default_doc();
        let mut buf = Vec::new();
        let mut ser = serde_json::Serializer::with_formatter(&mut buf, CanonicalFormatter::new());
        doc.serialize(&mut ser).expect("failed to serialize");
        let s = String::from_utf8(buf).expect("failed to convert bytes to string");
        insta::assert_yaml_snapshot!(s);
    }

    #[test]
    fn deserialize_embedded_keys() {
        let mut d = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        d.push("testdata/did_doc_embedded_keys.json");
        let file = File::open(d.as_path()).expect("failed to open test file");
        let doc: DidDocument = serde_json::from_reader(file).expect("failed to deserialize");
        insta::with_settings!(
            { sort_maps => true },
            {
                insta::assert_yaml_snapshot!(doc);
            }
        );
    }

    #[test]
    fn serialize_embedded_keys() {
        let doc = DidDocument {
            context: vec![
                Context {
                    url: Some("https://www.w3.org/ns/did/v1".to_string()),
                    url_map: None,
                },
                Context {
                    url: Some("https://w3id.org/security/suites/ed25519-2020/v1".to_string()),
                    url_map: None,
                },
            ],
            id: "did:example:123".to_string(),
            authentication: Some(vec![VmRelationship {
                key_id: None,
                verification_method: Some(VerificationMethod {
                    id: "did:example:123#z6MkecaLyHuYWkayBDLw5ihndj3T1m6zKTGqau3A51G7RBf3"
                        .to_string(),
                    type_: "Ed25519VerificationKey2020".to_string(),
                    controller: "did:example:123".to_string(),
                    public_key_multibase: Some(
                        "zAKJP3f7BD6W4iWEQ9jwndVTCBq8ua2Utt8EEjJ6Vxsf".to_string(),
                    ),
                    ..Default::default()
                }),
            }]),
            capability_invocation: Some(vec![VmRelationship {
                key_id: None,
                verification_method: Some(VerificationMethod {
                    id: "did:example:123#z6MkhdmzFu659ZJ4XKj31vtEDmjvsi5yDZG5L7Caz63oP39k"
                        .to_string(),
                    type_: "Ed25519VerificationKey2020".to_string(),
                    controller: "did:example:123".to_string(),
                    public_key_multibase: Some(
                        "z4BWwfeqdp1obQptLLMvPNgBw48p7og1ie6Hf9p5nTpNN".to_string(),
                    ),
                    ..Default::default()
                }),
            }]),
            capability_delegation: Some(vec![VmRelationship {
                key_id: None,
                verification_method: Some(VerificationMethod {
                    id: "did:example:123#z6Mkw94ByR26zMSkNdCUi6FNRsWnc2DFEeDXyBGJ5KTzSWyi"
                        .to_string(),
                    type_: "Ed25519VerificationKey2020".to_string(),
                    controller: "did:example:123".to_string(),
                    public_key_multibase: Some(
                        "zHgo9PAmfeoxHG8Mn2XHXamxnnSwPpkyBHAMNF3VyXJCL".to_string(),
                    ),
                    ..Default::default()
                }),
            }]),
            assertion_method: Some(vec![VmRelationship {
                key_id: None,
                verification_method: Some(VerificationMethod {
                    id: "did:example:123#z6MkiukuAuQAE8ozxvmahnQGzApvtW7KT5XXKfojjwbdEomY"
                        .to_string(),
                    type_: "Ed25519VerificationKey2020".to_string(),
                    controller: "did:example:123".to_string(),
                    public_key_multibase: Some(
                        "z5TVraf9itbKXrRvt2DSS95Gw4vqU3CHAdetoufdcKazA".to_string(),
                    ),
                    ..Default::default()
                }),
            }]),
            ..Default::default()
        };
        let mut buf = Vec::new();
        let mut ser = serde_json::Serializer::with_formatter(&mut buf, CanonicalFormatter::new());
        doc.serialize(&mut ser).expect("failed to serialize");
        let s = String::from_utf8(buf).expect("failed to convert bytes to string");
        insta::assert_yaml_snapshot!(s);
    }

    #[test]
    fn deserialized_referenced_keys() {
        let mut d = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        d.push("testdata/did_doc_keys_services.json");
        let file = File::open(d.as_path()).expect("failed to open test file");
        let doc: DidDocument = serde_json::from_reader(file).expect("failed to deserialize");
        insta::with_settings!(
            { sort_maps => true },
            {
                insta::assert_yaml_snapshot!(doc);
            }
        );
    }

    #[test]
    fn serialize_referenced_keys() {
        let doc = DidDocument {
            context: vec![
                Context {
                    url: Some("https://www.w3.org/ns/did/v1".to_string()),
                    url_map: None,
                },
                Context {
                    url: Some("https://w3id.org/security/suites/ed25519-2020/v1".to_string()),
                    url_map: None,
                },
            ],
            id: "did:example:123".to_string(),
            verification_method: Some(vec![VerificationMethod {
                id: "did:example:123#key-0".to_string(),
                type_: "JsonWebKey2020".to_string(),
                controller: "did:example:123".to_string(),
                public_key_jwk: Some(Jwk {
                    kty: "OKP".to_string(),
                    crv: Some("Ed25519".to_string()),
                    x: Some("VCpo2LMLhn6iWku8MKvSLg2ZAoC-nlOyPVQaO3FxVeQ".to_string()),
                    ..Default::default()
                }),
                ..Default::default()
            }]),
            service: Some(vec![
                Service {
                    id: "did:example:123#vcs".to_string(),
                    type_: vec!["VerifiableCredentialService".to_string()],
                    service_endpoint: vec![ServiceEndpoint {
                        url: Some("https://example.com/vc/".to_string()),
                        url_map: None,
                    }],
                },
                Service {
                    id: "did:example:123#wibble".to_string(),
                    type_: vec!["WibbleService".to_string()],
                    service_endpoint: vec![
                        ServiceEndpoint {
                            url: Some("https://example.com/wibbleThing/".to_string()),
                            url_map: None,
                        },
                        ServiceEndpoint {
                            url: None,
                            url_map: Some(HashMap::from_iter(vec![(
                                "mappyThing".to_string(),
                                vec!["https://example.com/mappyThing/".to_string()],
                            )])),
                        },
                    ],
                },
            ]),
            ..Default::default()
        };
        let mut buf = Vec::new();
        let mut ser = serde_json::Serializer::with_formatter(&mut buf, CanonicalFormatter::new());
        doc.serialize(&mut ser).expect("failed to serialize");
        let s = String::from_utf8(buf).expect("failed to convert bytes to string");
        insta::assert_yaml_snapshot!(s);
    }

    #[test]
    fn get_key() {
        let doc = default_doc();
        let key = doc
            .get_key(KeyPurpose::Authentication)
            .expect("failed to extract expected authentication key");
        assert_eq!(key.id, "371544b48d7d60d430c9c8b4af3745fa");
        let key = doc
            .get_key(KeyPurpose::AssertionMethod)
            .expect("failed to extract expected assertion method key");
        assert_eq!(
            key.controller,
            "did:ion:EiAscM5K0lfATv8GEqlR_RAVId0alzdcOgIRs-fBLXBWFA"
        );
    }
}
