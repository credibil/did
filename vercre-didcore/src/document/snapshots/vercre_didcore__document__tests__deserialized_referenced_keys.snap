---
source: vercre-didcore/src/document/mod.rs
assertion_line: 348
expression: doc
---
id: "did:example:123"
"@context":
  - "https://www.w3.org/ns/did/v1"
  - "https://w3id.org/security/suites/jws-2020/v1"
  - "@base": "https://example.com/keys/"
verificationMethod:
  - id: "did:example:123#key-0"
    type: JsonWebKey2020
    controller: "did:example:123"
    publicKeyJwk:
      kty: OKP
      crv: Ed25519
      x: VCpo2LMLhn6iWku8MKvSLg2ZAoC-nlOyPVQaO3FxVeQ
  - id: "did:example:123#key-1"
    type: JsonWebKey2020
    controller: "did:example:123"
    publicKeyJwk:
      kty: OKP
      crv: X25519
      x: pE_mG098rdQjY3MKK2D5SUQ6ZOEW3a6Z6T7Z4SgnzCE
  - id: "did:example:123#key-2"
    type: JsonWebKey2020
    controller: "did:example:123"
    publicKeyJwk:
      kty: EC
      crv: secp256k1
      x: Z4Y3NNOxv0J6tCgqOBFnHnaZhJF6LdulT7z8A-2D5_8
      y: i5a2NtJoUKXkLm6q8nOEu9WOkso1Ag6FTUT6k_LMnGk
  - id: "did:example:123#key-3"
    type: JsonWebKey2020
    controller: "did:example:123"
    publicKeyJwk:
      kty: EC
      crv: secp256k1
      x: U1V4TVZVMUpUa0ZVU1NBcU9CRm5IbmFaaEpGNkxkdWx
      y: i5a2NtJoUKXkLm6q8nOEu9WOkso1Ag6FTUT6k_LMnGk
  - id: "did:example:123#key-4"
    type: JsonWebKey2020
    controller: "did:example:123"
    publicKeyJwk:
      kty: EC
      crv: P-256
      x: Ums5WVgwRkRTVVFnU3k5c2xvZllMbEcwM3NPRW91ZzN
      y: nDQW6XZ7b_u2Sy9slofYLlG03sOEoug3I0aAPQ0exs4
  - id: "did:example:123#key-5"
    type: JsonWebKey2020
    controller: "did:example:123"
    publicKeyJwk:
      kty: EC
      crv: P-384
      x: VUZKSlUwMGdpSXplekRwODhzX2N4U1BYdHVYWUZsaXVDR25kZ1U0UXA4bDkxeHpE
      y: jq4QoAHKiIzezDp88s_cxSPXtuXYFliuCGndgU4Qp8l91xzD1spCmFIzQgVjqvcP
authentication:
  - "did:example:123#key-0"
assertionMethod:
  - "did:example:123#key-1"
  - "did:example:123#key-5"
service:
  - id: "did:example:123#vcs"
    type: VerifiableCredentialService
    serviceEndpoint: "https://example.com/vc/"
  - id: "did:example:123#wibble"
    type: WibbleService
    serviceEndpoint:
      origins:
        - "https://example.com/"
  - id: "did:example:123#wobble"
    type: WobbleService
    serviceEndpoint:
      - origins:
          - "https://example.com/"
          - "https://api.example.com/"
      - "https://example.com/wobble/"
      - "https://api.example.com/wobble/"

